### PR TITLE
drivers: spi: pl022: Add support DMA transfer

### PR DESCRIPTION
**Pull request by soburi**
Add supporting DMA-based transfer for PL022 SPI.
    
Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

---

This PR continues from #53892.
I Need to rebase after that merged.

### CODE REVIEW

**Review by teburd — APPROVED** [review]
Some minor nits, not -1 worthy

The test case covering the multitude of potential overlay+kconfig setups is interesting, it would be nice to replicate that not just for rpi_pico to ensure other SPI drivers are buildable in those scenarios.

File: drivers/spi/spi_pl022.c [hunk 1]
```diff
@@ -610,22 +883,40 @@ static int spi_pl022_init(const struct device *dev)
...
-	const struct spi_pl022_cfg *cfg = dev->config;
-
 	ret = pinctrl_apply_state(cfg->pincfg, PINCTRL_STATE_DEFAULT);
 	if (ret) {
 		LOG_ERR("Failed to apply pinctrl state");
 		return ret;
 	}
 #endif
 
+	if (cfg->dma_enabled) {
```
[the remark this answers is not in the document]
Comment (soburi) [reply drivers/spi/spi_pl022.c:898]: This section references a structure member whose existence depends on configuration.

```
#if defined(CONFIG_SPI_PL022_INTERRUPT)
	void (*irq_config)(const struct device *port);
#endif
#if defined(CONFIG_SPI_PL022_DMA)
	const struct spi_pl022_dma_config dma[NUM_OF_DIRECTION];
#endif
```

So this clause should switch at compile time.

File: drivers/spi/spi_pl022.c [hunk 2]
```diff
@@ -610,22 +883,40 @@ static int spi_pl022_init(const struct device *dev)
...
+			}
+
+			ret = dma_request_channel(cfg->dma[i].dev, &ch_filter);
+			if (ret < 0) {
+				LOG_ERR("dma_request_channel failed %d", ret);
+				return ret;
+			}
+		}
+#endif
+	} else {
```
[the remark this answers is not in the document]
Comment (soburi) [reply drivers/spi/spi_pl022.c:915]: Same as above.

File: drivers/spi/spi_pl022.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (teburd) [reply drivers/spi/spi_pl022.c:898]: IS_ENABLED will result in a constant and cause the compiler to sort out things appropriately

[the remark this answers is not in the document]
Comment (soburi) [reply drivers/spi/spi_pl022.c:898]: > IS_ENABLED will result in a constant and cause the compiler to sort out things appropriately

In my understanding, it is link time behavior.
We need to resolve all references at compile time.

(And I tried it, and it didn't work.)

**Comment by soburi** [comment]
@teburd 

> The test case covering the multitude of potential overlay+kconfig setups is interesting, it would be nice to replicate that not just for rpi_pico to ensure other SPI drivers are buildable in those scenarios.

I got it.
But this task is too complex to handle in this PR.
So I'll create a new PR to fix it.

**Review by carlescufi — APPROVED** [review]

**Pull request merged**